### PR TITLE
Migrate to new version

### LINKED ISSUE #9: Consider upgrading to simdjson 0.4.0
# Version 0.4 of simdjson is now available

Highlights

- Test coverage has been greatly improved and we have resolved many static-analysis warnings on different systems. 

New features:
- We added a fast (8GB/s) minifier that works directly on JSON strings.
- We added fast (10GB/s) UTF-8 validator that works directly on strings (any strings, including non-JSON). 
- The array and object elements have a constant-time size() method. 

Performance:
- Performance improvements to the API (type(), get<>()).
- The parse_many function (ndjson) has been entirely reworked. It now uses a single secondary thread instead of several new threads.
- We have introduced a faster UTF-8 validation algorithm (lookup3) for all kernels (ARM, x64 SSE, x64 AVX).

System support:
- C++11 support for older compilers and systems.
- FreeBSD support (and tests).
- We support the clang front-end compiler (clangcl) under Visual Studio.
- It is now possible to target ARM platforms under Visual Studio.
- The simdjson library will never abort or print to standard output/error.

# Version 0.3 of simdjson is now available

## Highlights

- **Multi-Document Parsing:** Read a bundle of JSON documents (ndjson) 2-4x faster than doing it  individually. [API docs](https://github.com/simdjson/simdjson/blob/master/doc/basics.md#newline-delimited-json-ndjson-and-json-lines) / [Design Details](https://github.com/simdjson/simdjson/blob/master/doc/parse_many.md)
- **Simplified API:** The API has been completely revamped for ease of use, including a new JSON  navigation API and fluent support for error code *and* exception styles of error handling with a  single API. [Docs](https://github.com/simdjson/simdjson/blob/master/doc/basics.md#the-basics-loading-and-parsing-json-documents)
- **Exact Float Parsing:** Now simdjson parses floats flawlessly *without* any performance loss (https://github.com/simdjson/simdjson/pull/558).
  [Blog Post](https://lemire.me/blog/2020/03/10/fast-float-parsing-in-practice/)
- **Even Faster:** The fastest parser got faster! With a [shiny new UTF-8 validator](https://github.com/simdjson/simdjson/pull/387)
  and meticulously refactored SIMD core, simdjson 0.3 is 15% faster than before, running at 2.5 GB/s   (where 0.2 ran at 2.2 GB/s).

## Minor Highlights

- Fallback implementation: simdjson now has a non-SIMD fallback implementation, and can run even on  very old 64-bit machines.
- Automatic allocation: as part of API simplification, the parser no longer has to be preallocated-it will adjust automatically when it encounters larger files.
- Runtime selection API: We've exposed simdjson's runtime CPU detection and implementation selection  as an API, so you can tell what implementation we detected and test with other implementations.
- Error handling your way: Whether you use exceptions or check error codes, simdjson lets you handle  errors in your style. APIs that can fail return simdjson_result<T>, letting you check the error  code before using the result. But if you are more comfortable with exceptions, skip the error code and cast straight to T, and exceptions will be thrown automatically if an error happens. Use the  same API either way!
- Error chaining: We also worked to keep non-exception error-handling short and sweet. Instead of  having to check the error code after every single operation, now you can *chain* JSON navigation calls like looking up an object field or array element, or casting to a string, so that you only  have to check the error code once at the very end.

### PR DESCRIPTION
**Pull request by sandrokeil**
Closes #9 #7 #6 

- [x] Fix deprecation warnings
- [x] Remove resource support (not sure how to handle it or if it makes sense at all)
- [x] Add support for key count
- [x] Add PHP benchmarks (`json_*` vs `simdjson_*`)
- [x] Throw exception in `key_value()` and `key_count()` if key is not available (catch c++ `simdjson::simdjson_error`)
- [x] Update readme 
- [x] Add Github action  CI

### CODE REVIEW

**Comment by crazyxman** [comment]
> **Don't merge**
> 
> Closes #9
> 
> Goal is to remove deprecation warnings.
@sandrokeil @WeaponMan 
Great, your contribution is a major improvement to "simdjson_php", I can't wait to merge it, thank you very much!

**Comment by sandrokeil** [comment]
@crazyxman It's ready for review.

**Comment by sandrokeil** [comment]
@crazyxman @WeaponMan Since I've added [using PHP exception](https://github.com/sandrokeil/simdjson_php/commit/48f033a5b91e9229b1a372de6d6968fce090eb8c), I got now the following error (*_ZNSt9exceptionD2Ev: symbol not found*) on compiling for Alpine Linux.

```
PHP Warning:  PHP Startup: Unable to load dynamic library 'simdjson.so' (tried: /tmp/simdjson/modules/simdjson.so (Error relocating /tmp/simdjson/modules/simdjson.so: _ZNSt9exceptionD2Ev: symbol not found), /tmp/simdjson/modules/simdjson.so.so (Error loading shared library /tmp/simdjson/modules/simdjson.so.so: No such file or directory)) in Unknown on line 0

PHP Warning:  PHP Startup: Unable to load dynamic library 'simdjson.so' (tried: /tmp/simdjson/modules/simdjson.so (Error relocating /tmp/simdjson/modules/simdjson.so: _ZNSt9exceptionD2Ev: symbol not found), /tmp/simdjson/modules/simdjson.so.so (Error loading shared library /tmp/simdjson/modules/simdjson.so.so: No such file or directory)) in Unknown on line 0
```

Any idea how to fix it or how to link it correctly? You can test it via Docker if you check out this branch run execute 

```
$ docker build -t simdjsontest -f docker_php7.4-alpine
```

**Comment by WeaponMan** [comment]
Last time I used alphine everything was compiled with `musl` libc as static binary, but I am not sure if alphine moved onto another libc in the meanwhile.
Anyway I would try instaling `libstdc++` and `libgcc` to see if it fixes the problem with the loading of the library, if it doesnt work you can try `ldd /tmp/simdjson/modules/simdjson.so` and it will show you whats missing.

**Comment by vnuk** [comment]
@sandrokeil I've tried building your branch on Ubuntu 18.04 with PHP installed from deb.sury.org - it looks like php version detection can't parse php version string 

```
 ./configure 
[snipped everything else]
checking PHP version... 7.3.20-1+ubuntu18.04.1+deb.sury.org+1
configure: error: not supported. Need a PHP version >= 7.3.0 (found 7.3.20-1+ubuntu18.04.1+deb.sury.org+1)
```

**Comment by sandrokeil** [comment]
@vnuk I restored the original PHP version check. Can you retry it with your version please?

@WeaponMan I've to add the stdc++ library.

@crazyxman It's ready for review again. [Tests are green](https://github.com/sandrokeil/simdjson_php/actions/runs/200495796). :tada:

**Comment by vnuk** [comment]
@sandrokeil compiles without issues, at first glance works without any issues 🎉

**Comment by crazyxman** [comment]
@sandrokeil 
```php
simdjson_decode("Invalid JSON string");
```
php-fpm: collapse
php-cli：
terminate called after throwing an instance of 'simdjson::simdjson_error'
  what():  The JSON document has an improper structure: missing or superfluous commas, braces, missing keys, etc.
Aborted。

expect：
Throw an exception

File: src/bindings.cpp [hunk 1]
```diff
@@ -31,7 +31,12 @@ simdjson::dom::element build_parsed_json_cust(const uint8_t *buf, size_t len, bo
     if (error) {
         std::cerr << "failure during memory allocation " << std::endl;
     } else {
-        doc = parser.parse(buf, len, realloc_if_needed);
+        try {
+            doc = parser.parse(buf, len, realloc_if_needed);
+        } catch (simdjson::simdjson_error &e) {
+            zend_throw_exception(spl_ce_RuntimeException, e.what(), 0 );
+            doc = parser.parse("{}"_padded);
```
Comment (sandrokeil) [inline src/bindings.cpp:38]: @crazyxman Good catch. The line `doc = parser.parse("{}"_padded);` is ugly, but I don't know how to solve it else. If this line is removed, then a `Segmentation fault (core dumped)` occurs. 

Any help would be appreciated.

**Comment by WeaponMan** [comment]
Take a look at https://github.com/WeaponMan/simdjson_php/commit/97d29005c12b403a698ebf6f82b3c87a5da361e3.
`build_parsed_json_cust` returns uninitialized memory if the parse fails so you getting segfault bcs of that.
I rewritten the error handling not to catch exception but pass around the `simdjson:error_code` instead.

File: src/bindings.cpp (hunk 1 above)
Comment (crazyxman) [reply src/bindings.cpp:38]: I have no good idea. 🙃

**Comment by sandrokeil** [comment]
@WeaponMan Can you bring your PR with the fix please. Looks promising.

**Pull request merged**